File: src/packages/control-ui/src/api/peer_acl_matrix.ts
export class PeerAclMatrix {
  private readonly acl = new Map<string, Map<string, boolean>>();
  private readonly defaultAllow: boolean;
  constructor(defaultAllow: boolean = true) {
    this.defaultAllow = defaultAllow;
  }

  public setRule(source: string, target: string, allowed: boolean): void {
    let row = this.acl.get(source);
    if (!row) {
      row = new Map();
      this.acl.set(source, row);
    }
    row.set(target, allowed);
  }

  public isAllowed(source: string, target: string): boolean {
    const row = this.acl.get(source);
    if (row && row.has(target)) {
      return row.get(target)!;
    }
    return this.defaultAllow;
  }
}

File: labs/mobile/donor-archive/tunnel/PeerAclMatrix.kt
// Copyright 2024-2026 LumiNet Authors. All Rights Reserved.
package com.luminet.android.tunnel

class PeerAclMatrix(private val defaultAllow: Boolean = true) {
    private val aclRules = mutableMapOf<String, MutableMap<String, Boolean>>()
    private val lock = Any()

    fun setRule(source: String, target: String, allowed: Boolean) = synchronized(lock) {
        aclRules.getOrPut(source) { mutableMapOf() }[target] = allowed
    }

    fun isAllowed(source: String, target: String): Boolean = synchronized(lock) {
        aclRules[source]?.get(target) ?: defaultAllow
    }
}
